File: modules/common/kv_db/kv_db.cc
/******************************************************************************
 * Copyright 2017 The Apollo Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *****************************************************************************/
#include "modules/common/kv_db/kv_db.h"

#include <sqlite3.h>

#include "gflags/gflags.h"

#include "cyber/common/log.h"
#include "modules/common/util/string_util.h"
#include "modules/common/util/util.h"

DEFINE_string(kv_db_path, "/apollo/data/kv_db.sqlite",
              "Path to Key-value DB file.");

namespace apollo {
namespace common {
namespace {
using apollo::common::util::StrCat;

// Self-maintained sqlite instance.
class SqliteWraper {
 public:
  static int Callback(void *data, int argc, char **argv, char **col_name) {
    if (data != nullptr) {
      std::string *data_str = static_cast<std::string *>(data);
      *data_str = argc > 0 ? argv[0] : "";
    }
    return 0;
  }

  SqliteWraper() {
    // Open DB.
    if (sqlite3_open(FLAGS_kv_db_path.c_str(), &db_) != 0) {
      AERROR << "Can't open Key-Value database: " << sqlite3_errmsg(db_);
      Release();
      return;
    }

    // Create table if it doesn't exist.
    static const char *kCreateTableSql =
        "CREATE TABLE IF NOT EXISTS key_value "
        "(key VARCHAR(128) PRIMARY KEY NOT NULL, value TEXT);";
    if (!SQL(kCreateTableSql)) {
      Release();
    }
  }

  ~SqliteWraper() { Release(); }

  bool SQL(const std::string &sql, std::string *value = nullptr) {
    AINFO << "Executing SQL: " << sql;
    if (db_ == nullptr) {
      AERROR << "DB is not open properly.";
      return false;
    }

    char *error = nullptr;
    if (sqlite3_exec(db_, sql.c_str(), Callback, value, &error) != SQLITE_OK) {
      AERROR << "Failed to execute SQL: " << error;
      sqlite3_free(error);
      return false;
    }
    return true;
  }

 private:
  void Release() {
    if (db_ != nullptr) {
      sqlite3_close(db_);
      db_ = nullptr;
    }
  }

  sqlite3 *db_ = nullptr;
};

}  // namespace

bool KVDB::Put(const std::string &key, const std::string &value) {
  SqliteWraper sqlite;
  return sqlite.SQL(
      StrCat("INSERT OR REPLACE INTO key_value (key, value) "
             "VALUES ('",
             key, "', '", value, "');"));
}

bool KVDB::Delete(const std::string &key) {
  SqliteWraper sqlite;
  return sqlite.SQL(StrCat("DELETE FROM key_value WHERE key='", key, "';"));
}

bool KVDB::Has(const std::string &key) {
  SqliteWraper sqlite;
  std::string value;
  const bool ret = sqlite.SQL(
      StrCat("SELECT value FROM key_value WHERE key='", key, "';"), &value);
  // Take empty field as non-exist.
  return ret && !value.empty();
}

std::string KVDB::Get(const std::string &key,
                      const std::string &default_value) {
  SqliteWraper sqlite;
  std::string value;
  const bool ret = sqlite.SQL(
      StrCat("SELECT value FROM key_value WHERE key='", key, "';"), &value);
  return (ret && !value.empty()) ? value : default_value;
}

}  // namespace common
}  // namespace apollo

File: modules/tools/common/kv_db.py
#!/usr/bin/env python2

###############################################################################
# Copyright 2017 The Apollo Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
###############################################################################
"""KV DB utils."""
import sys

import gflags
import plyvel

gflags.DEFINE_string("kv_db_path", "/apollo/data/kv_db", "Path to KV DB.")


class KVDB(object):
    """KV DB helper functions."""

    @staticmethod
    def Put(key, value):
        """Put an entry into DB."""
        db = plyvel.DB(gflags.FLAGS.kv_db_path, create_if_missing=True)
        db.put(key, value)
        db.close()

    @staticmethod
    def Delete(key):
        """Delete a key from DB."""
        db = plyvel.DB(gflags.FLAGS.kv_db_path, create_if_missing=True)
        db.delete(key)
        db.close()

    @staticmethod
    def Get(key):
        """Get value of a key from DB."""
        db = plyvel.DB(gflags.FLAGS.kv_db_path, create_if_missing=True)
        value = db.get(key)
        db.close()
        return value

    @staticmethod
    def List():
        """List values from DB."""
        db = plyvel.DB(gflags.FLAGS.kv_db_path, create_if_missing=True)
        for key, value in db:
            print(key, ":", value)
        db.close()


if __name__ == '__main__':
    def _help():
        print """Usage:
            {0} put <key> <value>
            {0} del <key>
            {0} get <key>
            {0} list
        """.format(sys.argv[0])
        sys.exit(0)

    if len(sys.argv) < 2:
        _help()

    gflags.FLAGS(sys.argv)
    op = sys.argv[1]
    if len(sys.argv) >= 3:
        key = sys.argv[2]
    if op == 'put':
        KVDB.Put(key, sys.argv[3]) if len(sys.argv) == 4 else _help()
    elif op == 'del':
        KVDB.Delete(key)
    elif op == 'get':
        print KVDB.Get(key)
    elif op == 'list':
        KVDB.List()
    else:
        _help()
